WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserFragment.java

```diff
@@ -0,0 +1,370 @@
+import org.wordpress.android.WordPress;
+import org.wordpress.android.ui.themes.ThemeBrowserAdapter.ScreenshotHolder;
+import org.wordpress.android.util.NetworkUtils;
+import org.wordpress.android.util.helpers.SwipeToRefreshHelper;
+import org.wordpress.android.util.helpers.SwipeToRefreshHelper.RefreshListener;
+import org.wordpress.android.util.widgets.CustomSwipeRefreshLayout;
+import org.wordpress.android.widgets.HeaderGridView;
+
+/**
+ * A fragment display the themes on a grid view.
+ */
+public class ThemeBrowserFragment extends Fragment implements RecyclerListener, AdapterView.OnItemSelectedListener, AbsListView.OnScrollListener {
+    public interface ThemeBrowserFragmentCallback {
+        void onActivateSelected(String themeId);
+        void onTryAndCustomizeSelected(String themeId);
+        void onViewSelected(String themeId);
+        void onDetailsSelected(String themeId);
+        void onSupportSelected(String themeId);
+        void onSearchClicked();
+    }
+
+    protected static final String BUNDLE_PAGE = "BUNDLE_PAGE";
+    protected static final int THEME_FILTER_ALL_INDEX = 0;
+    protected static final int THEME_FILTER_FREE_INDEX = 1;
+    protected static final int THEME_FILTER_PREMIUM_INDEX = 2;
+
+    protected SwipeToRefreshHelper mSwipeToRefreshHelper;
+    private String mCurrentThemeId;
+    private HeaderGridView mGridView;
+    private RelativeLayout mEmptyView;
+    private TextView mNoResultText;
+    private TextView mCurrentThemeTextView;
+    private ThemeBrowserAdapter mAdapter;
+    private Spinner mSpinner;
+    private ThemeBrowserFragmentCallback mCallback;
+    private int mPage = 1;
+    private boolean mShouldRefreshOnStart;
+    private TextView mEmptyTextView;
+
+    @Override
+    public void onAttach(Activity activity) {
+        super.onAttach(activity);
+
+        try {
+            mCallback = (ThemeBrowserFragmentCallback) activity;
+        } catch (ClassCastException e) {
+            throw new ClassCastException(activity.toString() + " must implement ThemeBrowserFragmentCallback");
+        }
+    }
+
+    @Override
+    public void onDetach() {
+        super.onDetach();
+        mCallback = null;
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.theme_browser_fragment, container, false);
+
+        setRetainInstance(true);
+        mNoResultText = (TextView) view.findViewById(R.id.theme_no_search_result_text);
+        mEmptyTextView = (TextView) view.findViewById(R.id.text_empty);
+        mEmptyView = (RelativeLayout) view.findViewById(R.id.empty_view);
+
+        configureGridView(inflater, view);
+        configureSwipeToRefresh(view);
+
+        return view;
+    }
+
+    @Override
+    public void onActivityCreated(Bundle savedInstanceState) {
+        ThemeBrowserActivity themeBrowserActivity = (ThemeBrowserActivity) getActivity();
```

Comment:
There are a ton of calls (and casts) to `getActivity()`. Probably better to just save a reference `onAttach` so you only have to cast once.